### PR TITLE
Add support for popular java tools

### PR DESCRIPTION
**Pull request by peterg79**
Implement support for clover code coverage and checkstyle

### CODE REVIEW

**Comment by Bachmann1234** [comment]
Neat! Ill review this in detail this weekend.

File: diff_cover/violationsreporters/java_violations_reporter.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        """
+        See super for args
+        """
+        super(CheckstyleXmlDriver, self).__init__(
+            'checkstyle',
+            ['java'],
+            ['java', '-jar', 'checkstyle-8.5-all.jar', '-c',
+             '/google_checks.xml']
+        )
+        self.command_to_check_install = ['java', '-cp', 'checkstyle-8.5-all.jar', 'com.puppycrawl.tools.checkstyle.Main', '-version']
```
Comment (Bachmann1234) [inline diff_cover/violationsreporters/java_violations_reporter.py]: Is there any way to have this less dependent on this specific version of checkstyle?

**Comment by Bachmann1234** [comment]
I like it, Im willing to loop past the .05% drop in coverage.

Can you update the readme to inform people about this. Make sure to include the versions of the supported tools.

**Comment by Bachmann1234** [comment]
Merging! I'll do a release sometime this week. Thanks!

**Comment by peterg79** [comment]
Thank you, this is a great tool :)

**Comment by Bachmann1234** [comment]
thanks! Released https://pypi.python.org/pypi/diff_cover/1.0.1

File: diff_cover/violationsreporters/java_violations_reporter.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    def _get_src_path_line_nodes(xml_document, src_path):
+        """
+        Return a list of nodes containing line information for `src_path`
+        in `xml_document`.
+
+        If file is not present in `xml_document`, return None
+        """
+        files = [file_tree
+                 for file_tree in xml_document.findall(".//file")
+                 if file_tree.get('path') == src_path
```
Comment (rishi05091989) [inline diff_cover/violationsreporters/java_violations_reporter.py]: @Bachmann1234 @peterg79 Before the equality check, we probably need to sanitize the path to unix path. Something like this [1]. Due to this clover reports generated under windows are not working with "GIT BASH"

[1] https://github.com/Bachmann1234/diff-cover/blob/da45062f066e94e0a65a84310b812c54cc423be9/diff_cover/violationsreporters/violations_reporter.py#L67

Comment (peterg79) [reply diff_cover/violationsreporters/java_violations_reporter.py]: @rishi05091989 Do you have some sample data I can use to test? I have no access to a windows computer these days.

Comment (rishi05091989) [reply diff_cover/violationsreporters/java_violations_reporter.py]: @peterg79 Thanks for your reply. The file tag in clover report looks like this,
```
<file path="C:\Users\guides_rishi\rishi_guides\guides\bundles\ds-guide-core\src\main\java\com\guide\themes\model\BreakpointInfo.java" name="BreakpointInfo.java">
```
 Also, is it possible to use the clover and cobertura XML together ? As of today, only one coverage report can be used at a time(mixed is not supported).

Comment (peterg79) [reply diff_cover/violationsreporters/java_violations_reporter.py]: > is it possible to use the clover and cobertura XML together
yes, with the latest release: https://github.com/Bachmann1234/diff-cover/pull/87

> need to sanitize the path to unix path
I'll take a look at it soon

**Pull request merged**